### ISSUE TITLE
CI: avoid GitHub Action cache over usage errors

### ISSUE THREAD

**Issue by cmur2** [camunda/camunda#18750]
## Description

In the monorepo we encounter intermittent build failures on jobs that use the GitHub Action cache (mostly indirectly via actions like `setup-java`, etc) with error messages like `Error: Cache upload failed because file read failed with EBADF: bad file descriptor, read` or `Warning: Failed to save: Cache service responded with 429 during upload chunk.` HTTP 429 status code indicates that "the user has sent too many requests in a given amount of time", i.e. rate limiting. A quick look at GitHub [docs](https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28#exceeding-the-rate-limit) says:

> If you exceed your primary rate limit, you will receive a 403 or 429 response, and the x-ratelimit-remaining header will be 0. You should not retry your request until after the time specified by the x-ratelimit-reset header.

Unfortunately the GHA toolkit used to upload cache does not yet [handle](https://github.com/actions/toolkit/issues/1589) HTTP 429 errors, which is really annoying.

We should seek a way to avoid these errors by adjusting CI workflows to avoid exceeding the limit.

### Observations

This happens during situations where the https://github.com/camunda/zeebe/actions/caches are dramatically overused and exceeding the limit: ![image](https://github.com/camunda/zeebe/assets/800933/e2620882-1840-487d-bac2-17550b54448c)

Experience with this limit shows that over usage above 10G does not immediately lead to failures but if it gets too much, the GitHub API will prevent uploading data to the cache.

This failure is more likely in the monorepo because several CI workflows taken from other repositories have not been adjusted yet to the fact that all of them share the common 10G limit now. The cache usage details can be retrieved via the [API](https://docs.github.com/en/rest/actions/cache?apiVersion=2022-11-28#get-github-actions-cache-usage-for-a-repository).

Previous reports of this problem:

* https://camunda.slack.com/archives/C06HTSPD5AP/p1713969515800199
* https://camunda.slack.com/archives/C071KP5BTHB/p1716389476038289 with https://github.com/camunda/zeebe/actions/runs/9193202260/job/25283661896?pr=18647
* https://camunda.slack.com/archives/C071KP5BTHB/p1718292962257739 (problem mainly triggered by a change related to https://github.com/camunda/team-infrastructure/issues/517)

### Known Workarounds

Retry failed workflows manually after waiting for a short time. GitHub automatically prunes the GitHub Actions caches when exceeding the limit so impacted CI workflows will succeed again afterwards.

**Comment by mustafadagher** [comment]
I've not been able to merge my [PR](https://github.com/camunda/camunda/pull/19340#issuecomment-2165976749) today for with errors in the merge CI for the same reason. In 4 merge tries it failed for all of them. I'm sure others are suffering from the same too. 
Is someone already looking into this?

**Comment by cmur2** [comment]
> Is someone already looking into this?

This problem was noticed by others and [reported via Slack](https://camunda.slack.com/archives/C071KP5BTHB/p1718292962257739) which is the preferred method to report problems since more people can look at it and react quickly.

**Comment by cmur2** [comment]
To address this I want to achieve metrics collection via https://github.com/camunda/team-infrastructure/issues/656 and then reduce cache over-usage based on that data.

**Comment by cmur2** [comment]
> Likely we would greatly reduce our cache usage if we centralized things. As it is, most PRs now create gigabytes of cache, as only Zeebe is set up to not create caches for PRs.

Very good points all around, I think we should come up with a GHA/CI caching strategy for the monorepo. Some ideas were:

1. only populate GHA cache from `main` and `stable` branches (by default `actions/cache` also does this from PRs → doesn't scale well for many PRs)
2. use same cache key for Operate/Zeebe/etc to share central caches
3. unlike 1. also create caches for PRs to speed up builds there as well but [force remove them on merge to avoid overusage](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/caching-dependencies-to-speed-up-workflows#force-deleting-cache-entries) (approach from GH)

**Comment by npepinpe** [comment]
I think 1 is a no brainer, and we should 100% do it. I could find someone from our team to do it to get more experience with CI :+1: It hopefully wouldn't be too big of a deal since we already have it set up for the Zeebe CI, so it's just applying it across the board (with your guidance :upside_down_face:).

For 2, I think now that everything is in the mono repo, it should work right? It does mean having to build the complete mono repo before caching, but then it's fine.

I can't speak for 3 - I suspect it won't matter much, as the changes to the dependency cache in PRs is likely not much, but this is an educated guess.

4 also makes sense, we've had more trouble with the Docker cache than anything, and it's not the bottleneck imo.

As for 5 and 6, I can't say - 6 seems to make sense for me as well, assuming performance is similar. I suspect with the monorepo and more supported versions we would need more than 10GB anyway of cache.

**Comment by cmur2** [comment]
> It hopefully wouldn't be too big of a deal since we already have it set up for the Zeebe CI, so it's just applying it across the board (with your guidance 🙃).

I noticed that nobody except Zeebe CI is using Maven Wrapper which might make this a bit more effort then since https://github.com/camunda/camunda/blob/ci-disable-docker-cache-in-prs/.github/actions/setup-zeebe/action.yml#L113-L180 seems to be tailored carefully to work with Maven Wrapper (in general probably makes sense to move that to a new composite action for reuse).

> I can't speak for 3 - I suspect it won't matter much, as the changes to the dependency cache in PRs is likely not much, but this is an educated guess.

3 would save some effort since we could continue to create caches on PRs but especially for short-lived PRs merged during 1 day (like dependency upgrades) the caches wouldn't idle around until the next nightly pruning → reduces pressure.

> 4 also makes sense, we've had more trouble with the Docker cache than anything, and it's not the bottleneck imo.

My assessment as well, only users are https://github.com/camunda/camunda/pull/21151 and [zeebe-benchmark.yml](https://github.com/camunda/camunda/blob/main/.github/workflows/zeebe-benchmark.yml) (not sure exactly why, can probably remove as well.)

**Comment by cmur2** [comment]
Unintentionally closed in https://github.com/camunda/camunda/pull/21182, still an open topic in general. Current status:

* https://github.com/camunda/camunda/pull/21151 + https://github.com/camunda/camunda/pull/21182 got us outside of dangerous usage levels
* POC needed to see how big impact of disabling caches for PRs is on workflow execution times:
    * if negligible then part of the strategy will become "1. only populate GHA cache from main and stable branches"
    * if it impacts UX too much, see if [GitHubs approach is better](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/caching-dependencies-to-speed-up-workflows#force-deleting-cache-entries)
* document strategy after POC and create action items to implement it

**Comment by lenaschoenburg** [comment]
> POC needed to see how big impact of disabling caches for PRs is on workflow execution times

Do we _really_ need a POC for this? Average PRs will not introduce more than one or two new dependencies and downloading those will not make a noticeable impact.

Additionally, the Zeebe CI has been running successfully with that approach for quite a while now.

**Comment by cmur2** [comment]
> Average PRs will not introduce more than one or two new dependencies and downloading those will not make a noticeable impact.

Fair point.

> Additionally, the Zeebe CI has been running successfully with that approach for quite a while now.

I'm not sure on the effects for the Yarn/Node ecosystem since Zeebe does not involve any frontend dependencies. The scope of this ticket is all the caching so backend, frontend, Docker, build tools, etc.

**Comment by cmur2** [comment]
I started to document what we said here [in the wiki](https://github.com/camunda/camunda/wiki/CI-&-Automation#github-actions-cache) and will create tickets based on that.

> I could find someone from our team to do it to get more experience with CI 👍 It hopefully wouldn't be too big of a deal since we already have it set up for the Zeebe CI, so it's just applying it across the board (with your guidance 🙃).

@npepinpe does this proposal still hold? can also discuss via DM once I have a ticket for that

**Comment by npepinpe** [comment]
Yes, we've discussed it with @rodrigo-lourenco-lopes To be clear we would need to set a timeline as we've discussed him spendong 10-15% of his time on CI, no more, so if it's critical then that wouldn't be a good fit.

**Comment by cmur2** [comment]
Related problem: https://camunda.slack.com/archives/C071KP5BTHB/p1726482584788839
